File: 2023/04/silver.go
package main

import (
	"fmt"
	"os"
	"regexp"
	"slices"
	"strconv"
	"strings"
)

type card struct {
	id           int
	leftNumbers  []int
	rightNumbers []int
}

func check(err error) {
	if err != nil {
		panic(err)
	}
}
func main() {
	bytes, err := os.ReadFile("input.txt")
	check(err)
	input := string(bytes)

	cards := formatInput(input)

	sum := 0
	for _, card := range cards {
		sum += card.getValue()
	}
	fmt.Println(sum)
}

func formatInput(input string) (cards []card) {
	lines := strings.Split(input, "\n")

	cardRegex := regexp.MustCompile("Card +(\\d+): +(.+) +[|] +(.+)")

	for _, line := range lines {
		if len(line) == 0 {
			continue
		}
		result := cardRegex.FindStringSubmatch(line)
		id := result[1]
		left := result[2]
		right := result[3]

		var leftNumbers []int
		var rightNumbers []int
		for _, value := range strings.Split(left, " ") {
			if len(value) == 0 {
				continue
			}
			number, _ := strconv.Atoi(value)
			leftNumbers = append(leftNumbers, number)
		}
		for _, value := range strings.Split(right, " ") {
			if len(value) == 0 {
				continue
			}
			number, _ := strconv.Atoi(value)
			rightNumbers = append(rightNumbers, number)
		}

		idNumber, _ := strconv.Atoi(id)
		cards = append(cards, card{
			id:           idNumber,
			leftNumbers:  leftNumbers,
			rightNumbers: rightNumbers,
		})

	}

	return cards
}

func (theCard card) getValue() (value int) {
	for _, number := range theCard.rightNumbers {
		if slices.Contains(theCard.leftNumbers, number) {
			if value == 0 {
				value = 1
			} else {
				value *= 2
			}
		}
	}
	return value
}

File: 2024/01/silver.rb
def main
  raw_input = File.read("input.txt")
  left_numbers = []
  right_numbers = []
  raw_input.split("\n").each do |line|
    parts = line.split("   ")
    left  = parts[0]
    right = parts[1]
    insert_in_order(left_numbers, left.to_i)
    insert_in_order(right_numbers, right.to_i)
  end
  sum_diff = 0
  left_numbers.each_with_index do |item,index|
    sum_diff += diff(item, right_numbers[index])
  end
  puts sum_diff
end

def insert_in_order(array, number)
  array.each_with_index do |item, index|
    if number < item
      array[index,0] = number
      return
    end
  end
  array.push number
end

def diff(a, b)
  (a - b).abs
end

main